### PR TITLE
Site Editor: Remove the details pages

### PR DESCRIPTION
**Pull request by youknowriad**
Related to #59689 

## What?

This PR is kind of the latest step of the unification of the "details panels" and the "inspector panel". Now that all/most of these panels unification is done, we can actually remove the "details pages" entirely.

That said, it's not without challenges when it comes to "routing" and "urls": Pages in the site editor can be loaded using two URLs today:

 - postType=page&postId=someID  This URL loads the page with the "details" panel open
 - path=%2Fpage&postId=270 This URL loads the page without the "details" panel (in list view)
 
As we're removing the details panel step, we need to unify. We have to pick one URL and redirect the other. I decided to keep the former as it's more friendly. 

This also means a change in all places where we link to these pages...

That said, it brings a lot of consistency and simplicity to our pages IMO, so I think it's a good change anyway.

## Testing Instructions

This is a very impactful change in the site editor, we need to test all the flows properly. E2E tests should help us catch some but probably not all the potential regressions.

### CODE REVIEW

File: packages/edit-site/src/components/add-new-pattern/index.js
Comment (youknowriad) [inline packages/edit-site/src/components/add-new-pattern/index.js:172]: For the distinction between patterns and template parts, I just used the postType and remove the categoryType from the url.

File: packages/edit-site/src/components/layout/router.js
Comment (youknowriad) [inline packages/edit-site/src/components/layout/router.js:23]: These components have been removed entirely, I still need to check if they have dependencies that are only used in the components themselves to avoid leftovers.

File: packages/edit-site/src/components/layout/router.js
Comment (youknowriad) [inline packages/edit-site/src/components/layout/router.js:33]: This hook centralizes the redirection of the old paths to the new paths. 

@jsnajdr I noticed a bug though, if I type a new url while the site editor is loaded, the sidebar of the site editor doesn't update properly but if I refresh the page, it does render the right sidebar. (So the redirect was successful but something is not re-rendering properly as the redirect happens)

File: packages/edit-site/src/components/sidebar-dataviews/index.js
Comment (youknowriad) [inline packages/edit-site/src/components/sidebar-dataviews/index.js:19]: I think we should unify between "activeView" in the pages page and "categoryId" in the patterns page.

File: packages/edit-site/src/components/sidebar-navigation-item/index.js
Comment (youknowriad) [inline packages/edit-site/src/components/sidebar-navigation-item/index.js:32]: Previously the path was a string and being used as a unique id for the "focus handling". I added an explicit uid to address the fact that now "params" is an object (I didn't want to use the postType directly as I wanted to support any link) cc @jsnajdr

File: packages/edit-site/src/hooks/commands/use-common-commands.js
Comment (youknowriad) [inline packages/edit-site/src/hooks/commands/use-common-commands.js:180]: This function was already broken before this PR. I changed the logic a bit to navigate to the "global styles" pages only when no active post/template is selected (no postId in the url)

File: packages/edit-site/src/components/sidebar-navigation-screen-template/home-template-details.js
Comment (ntsekouras) [inline packages/edit-site/src/components/sidebar-navigation-screen-template/home-template-details.js:4]: Can you keep this around for a bit? We'll have to move them to inspector controls and it might be helpful to have the existing code. I'll see to open a PR for the next days.

**Comment by jasmussen** [comment]
It's the right time to do it, so I'm happy to move forward. It certainly makes the navigation much clearer:

flow

File: packages/edit-site/src/components/add-new-pattern/index.js
Comment (ntsekouras) [reply packages/edit-site/src/components/add-new-pattern/index.js:172]: I'm not sure if this is the culprit of what I'm seeing, but template parts doesn't seem to work well..

https://github.com/WordPress/gutenberg/assets/16275880/8b2a3eac-046f-4a2e-9332-017cb9cec0db

File: packages/edit-site/src/components/sidebar-dataviews/index.js
Comment (ntsekouras) [reply packages/edit-site/src/components/sidebar-dataviews/index.js:19]: Maybe we should do that when we actually use views in patterns page to avoid creating extra confusion with nomenclature.

File: packages/edit-site/src/components/layout/router.js
Comment (ntsekouras) [reply packages/edit-site/src/components/layout/router.js:33]: There is also the case with `http://localhost:8888/wp-admin/site-editor.php?postId=twentytwentyfour%2F%2Fwp-custom-template-sss&postType=wp_template`. Previously we would render the page with the details panel, and now it goes to the list. The params are effective visually in `pages` where we have the `list` view as default, but not in templates etc.. Should we redirect to `edit` mode in these cases or remove the params altogether? 🤔

**Comment by ntsekouras** [comment]
I did a first round of testing and seems to work quite well 🚀 . There are things to iron out, but lots of testing will be needed here.

File: packages/edit-site/src/components/layout/router.js
Comment (youknowriad) [reply packages/edit-site/src/components/layout/router.js:33]: We have a couple options: we can switch the templates to list view as well or redirect to "edit". Leaning towards edit for now.

File: packages/edit-site/src/components/sidebar-dataviews/index.js
Comment (youknowriad) [reply packages/edit-site/src/components/sidebar-dataviews/index.js:19]: Isn't it just the "default views" of patterns though, just like we have "default views" for pages (regardless of whether you can create views)

File: packages/edit-site/src/components/sidebar-navigation-screen-template/home-template-details.js
Comment (youknowriad) [reply packages/edit-site/src/components/sidebar-navigation-screen-template/home-template-details.js:4]: If this should block the PR, it's fine we can wait but if it shouldn't, we can always access the code in the old branches.

File: packages/edit-site/src/components/layout/router.js
Comment (youknowriad) [reply packages/edit-site/src/components/layout/router.js:33]: On a second though, I'm not sure we should redirect here to another page here. This URL is valid even in the new url scheme. What we're missing is solving the bug of the "controlled selection". In other words, even in the table block, the "row" should appear as selected.

File: packages/edit-site/src/components/sidebar-dataviews/index.js
Comment (ntsekouras) [reply packages/edit-site/src/components/sidebar-dataviews/index.js:19]: Semantically yes, technically no. With nomenclature confusion I meant mostly towards devs.

File: packages/edit-site/src/components/sidebar-navigation-screen-template/home-template-details.js
Comment (ntsekouras) [reply packages/edit-site/src/components/sidebar-navigation-screen-template/home-template-details.js:4]: Of course it shouldn't block this PR. I'll try to open a PR for that asap though.

**Comment by youknowriad** [comment]
@jasmussen @jameskoster Here's an interesting challenge that we might not have thought about:

 - If you go to the "pages" page
 - navigate using pagination (page 2)
 - click on the preview to open a random page
 - click the site icon to go back.
 - Ideally, you'd expect to be back to the same "page 2" with the random page that you opened selected right?

This we can't implement at the moment and to be honest, I'm not sure what the ideal scenario for all cases is, knowing that:

 - You can insert a random editor by sharing a URL (so no history)
 - You can access the editor from list view, table view or any random page (like using the command palette)

I do wonder whether we should actually just simplify the "site icon" to mean just go back to the listing of the pages (kind of like classic wp-admin or post editor)

**Comment by jameskoster** [comment]
Yeah there are many similar flows; Adding filters, changing layout, toggling fields, sorting, switching to a different view (Drafts). In all of these cases you'd probably expect the W button to return you to exactly where you came from.

The W button is a complex beast, and I don't suppose this is the PR to figure out all it's behaviors. The suggestion (always returns to list) seems okay for now.

File: packages/edit-site/src/components/layout/router.js
Comment (jsnajdr) [reply packages/edit-site/src/components/layout/router.js:33]: > There is also the case with `postId=...&postType=wp_template`

This URL, if it doesn't have `canvas=edit`, now means "show the list of posts of type `wp_template`, and select the `postId` there if you can." For `postType=page` it actually highlights something, but for other types the list UI doesn't use the `postId` parameter. But it could.

> if I type a new url while the site editor is loaded, the sidebar of the site editor doesn't update properly

@youknowriad This is weird, can you post a video, or specify which URL you were on and which new URL you typed?

File: packages/edit-site/src/components/layout/router.js
Comment (jsnajdr) [inline packages/edit-site/src/components/layout/router.js:84]: Seems the only `path` remaining is `/wp_global_styles`, right? It should be the first `if` in this function. If I go to an ambiguous URL like `path=/wp_global_styles&postType=wp_template`, the `path` should win.

Also, one gap in the current UI is what kind of sidebar to show for post types that are not explicitly handled in this function. If I go to `postType=post&postId=42`, it will show the right post (because that's handled by `useInitEditedEntityFromUrl`), but will show the root sidebar. This problem is not new though, the old version would show you the Navigations sidebar, also not great.

Comment (youknowriad) [reply packages/edit-site/src/components/layout/router.js:84]: > If I go to postType=post&postId=42, it will show the right post (because that's handled by useInitEditedEntityFromUrl), but will show the root sidebar. This problem is not new though, the old version would show you the Navigations sidebar, also not great.

Interesting, I think for now we should forbid this and not support other post types in the site editor.

Comment (jsnajdr) [reply packages/edit-site/src/components/layout/router.js:84]: By the way, if we want to forbid editing posts and other post types in the site editor, what is the idea behind supporting page editing there? What is so special about pages that we need a second UI for them in addition to `edit.php?post_type=page`?

Comment (youknowriad) [reply packages/edit-site/src/components/layout/router.js:84]: There's nothing special about it. Maybe they're a bit closer to "design/appearance" than the rest of post types. But ultimately, they don't belong there, it's just a tactical move as we iterate towards the redesign of WP-Admin. More details here https://github.com/WordPress/gutenberg/issues/55238#issuecomment-2118253846

Comment (jsnajdr) [reply packages/edit-site/src/components/layout/router.js:84]: Thanks, that explains a lot 🙂

Comment (youknowriad) [reply packages/edit-site/src/components/layout/router.js:84]: For now, I'm treating all non-supported post types the same and just rendering the home page of the site editor.

File: packages/edit-site/src/components/layout/router.js
Comment (youknowriad) [reply packages/edit-site/src/components/layout/router.js:33]: @jsnajdr Actually, it's broken for me even when loading a page in a new tab:

```
http://localhost:8888/wp-admin/site-editor.php?path=%2Fwp_template
```

Comment (youknowriad) [reply packages/edit-site/src/components/layout/router.js:33]: Ok I found the issue, it's a race condition between the effect in The RoutesProvider and the effect here. child effects run before parents. I'll see if I can come up with the correct fix.

Comment (youknowriad) [reply packages/edit-site/src/components/layout/router.js:33]: Solved here [2af8009](https://github.com/WordPress/gutenberg/pull/61741/commits/2af80090922e45bafb9026ada695bd1a751eb27c)

File: packages/edit-site/src/components/add-new-pattern/index.js
Comment (youknowriad) [reply packages/edit-site/src/components/add-new-pattern/index.js:172]: This is fixed.

**Comment by draganescu** [comment]
Testing this I found that accessing a couple of pages and then visiting the trash section crashes the browser:


<details><summary>Video</summary>
<p>


https://github.com/WordPress/gutenberg/assets/107534/d5ac03fe-22d5-4e12-8aac-5bba7bda0c6c



</p>
</details> 



Also on small viewports there seems to be no way to go back to the root of the site editor sidebar navigation:

_EDIT_ - this issue seems to also be in trunk.

<details><summary>Video</summary>
<p>

https://github.com/WordPress/gutenberg/assets/107534/d8a31834-9e48-4254-8c0d-35ff8ba48fe7


</p>
</details>

**Comment by youknowriad** [comment]
@draganescu the crash has been fixed in [62b7061](https://github.com/WordPress/gutenberg/pull/61741/commits/62b7061d807b1e4660b7dd336e858912fbd24edd)

**Comment by youknowriad** [comment]
Should we move forward here, what's left?

File: packages/router/src/router.js
Comment (jsnajdr) [inline packages/router/src/router.js:39]: I'm slightly confused by this 🙂 If we `clear()` the `locationCache` all the time like here, it will ever remember only the last value, and the map will always have just one record, right? That's a common way of memoization, but usually not implemented with a `Map`.

Also, it works only as long as all our routing is happening inside the query params. When you start routing also with a real path, and the location changes from:
```
{ path: '/site-editor', search: 'postId=1' }
```
to
```
{ path '/edit', search: 'postId=1' }
```
then `getLocationWithParams` will return the old `/site-editor` location, because that's what's been memoized under the `postId=1` key.

Comment (youknowriad) [reply packages/router/src/router.js:39]: you're right, for the path, we actually don't care about it but initially when I tried using weak map and use the entire location object as a key, the component was re-rendering way too much (location didn't seem stable for some reason, see this commit [62b7061](https://github.com/WordPress/gutenberg/pull/61741/commits/62b7061d807b1e4660b7dd336e858912fbd24edd) 

Happy to use an alternative implementation, anything you have in mind?

**Review by ellatrix — APPROVED** [review]
I like this a lot, also agree that the postType url is better.
Let's get this in asap so it gets some testing in trunk before Beta 1.

File: packages/edit-site/src/components/layout/router.js
Comment (ntsekouras) [inline packages/edit-site/src/components/layout/router.js]: Nit: the name is not correct here. Could we just use one destructuring instead of two to simplify things a bit?

Comment (youknowriad) [reply packages/edit-site/src/components/layout/router.js]: Good catch. 7fec893

**Review by draganescu — DISMISSED** [review]
Somehow this PR undoes #61275  and makes navigating to a navigation also change the content in the frame.

File: packages/edit-site/src/components/sidebar-navigation-item/index.js
Comment (draganescu) [reply packages/edit-site/src/components/sidebar-navigation-item/index.js:32]: Is this `uid` data? Would a `data-` attribute make more sense? I was wondering if it's an id why not use the id attribute then?

**Comment by youknowriad** [comment]
@draganescu Can you explain more, in my testing it's behaving exactly like shown in the video of the linked PR, so I'm not entirely sure what to expect

File: packages/edit-site/src/components/layout/router.js
Comment (jsnajdr) [reply packages/edit-site/src/components/layout/router.js:84]: > the only path remaining is `/wp_global_styles`, right? It should be the first `if` in this function.

This feedback is I believe still valid, but unaddressed. 🙂

**Comment by draganescu** [comment]
<details><summary>In this PR</summary>
<p>

https://github.com/WordPress/gutenberg/assets/107534/18409818-4c62-4c71-84b2-05bf75a96364

</p>
</details> 


<details><summary>In trunk</summary>
<p>

https://github.com/WordPress/gutenberg/assets/107534/3f847673-7436-499c-83f9-d20139a8fda8

</p>
</details>

**Comment by youknowriad** [comment]
@draganescu  The behavior in trunk is not consistent. When you have just a single menu, you can see that going back from editor to navigation you pass by the state of the "current navigation menu" selected in the frame too. the behavior in the PR is consistent though. So I think I prefer the behavior on this PR personally.

**Comment by draganescu** [comment]
No, the inconsistency is a bug that should be fixed, but the fix is not to undo the PR. That PR had plenty of conversation around why the single post type editing for navigation menus should not be a default experience.

**Comment by ntsekouras** [comment]
Personally I also prefer the routing of menus here and think we should avoid the inconsistency..

File: packages/edit-site/src/components/layout/router.js
Comment (ntsekouras) [inline packages/edit-site/src/components/layout/router.js:178]: Nit: not from this PR but the only different thing we return in `if ( postType === NAVIGATION_POST_TYPE ) {` is the  backpath and we could apply the `if ( postId ) {` check there.

**Review by ntsekouras — COMMENTED** [review]
This looks good and tests well for me.

**Review by jasmussen — APPROVED** [review]
This is working well enough for me. 

Andrei notes that this undoes work in #61275. That work came from a place to improve navigation, which is valid: it's clear that the navigation section is not in a great place. Nevertheless there's an issue where you can still edit that navigation in isolation, then press "Back", and land in a pretty incongruent state:

nav

It's not clear that aspect was accounted for in 61275, and to me it doesn't seem an acceptable compromise. Not sure the best path forward, happy to defer to others. But at the moment, the behavior for navigation in this branch, and consequent in 6.5 which it is similar to, seems better than what's shipping in trunk.

**Comment by richtabor** [comment]
>  Nevertheless there's an issue where you can still edit that navigation in isolation, then press "Back", and land in a pretty incongruent state:
> It's not clear that aspect was accounted for in 61275, and to me it doesn't seem an acceptable compromise. 

I'm pretty sure the "Edit" dropdown to engage edit mode was intentionally left in, as a method for getting to the edit view of the navigation menu. 

Both are not entirely ideal, though we don't have a plan together for making the navigation edit view decent. 

The intention behind #61275 was to reduce the likelihood of ending up in the navigation edit view, where it's wholly unfinished and confusing—as it very well may not represent the way that particular navigation block looks/is placed. 

That's why the inconsistency was introduced; to prevent previewing the navigation in a confusing experience.

File: packages/router/src/router.js
Comment (jsnajdr) [reply packages/router/src/router.js:39]: > Happy to use an alternative implementation, anything you have in mind?

For now you can leave it as it is, I'm working on a parallel PR that moves `getLocationWithParams` to the `history` object, there are a few other things that will be fixed by this.

> initially when I tried using weak map and use the entire location object as a key, the component was re-rendering way too much

This is strange because the `history` package returns stable and frozen `location` objects. I'll try out the commit with weakmap.

Comment (jsnajdr) [reply packages/router/src/router.js:39]: Alternative implementation proposed here: #61823.

File: packages/edit-site/src/components/layout/router.js
Comment (youknowriad) [reply packages/edit-site/src/components/layout/router.js:178]: The sidebar component used is different as well.

File: packages/router/src/router.js
Comment (youknowriad) [reply packages/router/src/router.js:39]: Left a comment there, it seems your implementation suffers from the same issue I had here with weakmap. Going to merge this PR but happy to iterate on this in your PR.

**Comment by ellatrix** [comment]
Looks like this PR caused a performance regression for site editor first block load. Any idea what it could be?

**Comment by Mamaduka** [comment]
I think this has changed the "Edit site" links behavior in the WP admin bar. It should open the currently rendered template in "edit" mode; now, it opens the template list view.

## Screencast

https://github.com/WordPress/gutenberg/assets/240569/2b4be013-d974-4fc0-b882-d97d8baf04a7

**Comment by youknowriad** [comment]
@ellatrix I change the url to check for the "first block" metric to canvas "edit" in this PR. The reason is that this url exists similarly in both this version and the old version of the site editor. I'm guessing that the "editor frame" is probably slower to load recently which might have caused the jump in performance. So I don't really think there's a performance regression in this PR per se but I'm not sure what to do here. If I restore the previous url, we'll be testing two different things: the details page that was removed with the page list (with the page visible in the frame). 

@Mamaduka Good catch, I'll see if I can fix this.

**Pull request merged**
